File: services/requests.ts
import { activeServices } from '@/data/active-services';
import { claimedPerks } from '@/data/claimed-perks';
import { requests } from '@/data/requests';
import { Request, RequestItem } from '@/types';

import { delay } from './mock-delay';

let requestCounter = requests.length;

type SubmitRequestPayload = {
  employeeId: string;
  employeeName: string;
  type: 'single' | 'bundle';
  items: RequestItem[];
  totalAll: number;
};

export async function submitRequest(payload: SubmitRequestPayload): Promise<Request> {
  requestCounter += 1;
  const request: Request = {
    id: `req-${requestCounter}`,
    ...payload,
    status: 'pending',
    createdAt: new Date().toISOString(),
  };
  requests.unshift(request);

  request.items.forEach((item) => {
    activeServices.unshift({
      id: `service-${item.perkId}-${request.id}`,
      employeeId: request.employeeId,
      requestId: request.id,
      title: item.title,
      providerName: item.providerName,
      priceAll: item.discountedPriceAll,
      status: 'pending',
      requestedAt: request.createdAt,
    });
  });

  return delay(request);
}

export async function getActiveRequests(): Promise<Request[]> {
  return delay(requests.filter((request) => request.status === 'pending'));
}

export async function getAllRequests(): Promise<Request[]> {
  return delay(requests);
}

type PaymentMethod = 'card' | 'paypal';

export async function approveRequest(requestId: string, paymentMethod: PaymentMethod): Promise<Request> {
  const request = requests.find((item) => item.id === requestId);
  if (!request) throw new Error('Request not found');
  request.status = 'approved';
  request.paidAt = new Date().toISOString();
  request.paymentMethod = paymentMethod;

  request.items.forEach((item) => {
    claimedPerks.unshift({
      id: `claim-${item.perkId}-${request.id}`,
      employeeId: request.employeeId,
      title: item.title,
      providerName: item.providerName,
      priceAll: item.discountedPriceAll,
      claimedAt: request.paidAt!,
    });
  });

  activeServices
    .filter((service) => service.requestId === request.id)
    .forEach((service) => {
      service.status = 'active';
    });

  return delay(request);
}

export async function declineRequest(requestId: string): Promise<Request> {
  const request = requests.find((item) => item.id === requestId);
  if (!request) throw new Error('Request not found');
  request.status = 'declined';

  for (let i = activeServices.length - 1; i >= 0; i -= 1) {
    if (activeServices[i].requestId === request.id) {
      activeServices.splice(i, 1);
    }
  }

  return delay(request);
}

export async function getClaimedPerks(employeeId: string) {
  return delay(claimedPerks.filter((claim) => claim.employeeId === employeeId));
}

File: backend/app/api/routers/requests.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_employee, get_current_manager
from app.db.session import get_db
from app.models.active_service import ActiveService
from app.models.employee import Employee
from app.models.redeemed_history import RedeemedHistory
from app.models.request import Request
from app.models.request_item import RequestItem
from app.models.service import Service
from app.schemas.request import RequestApprove, RequestCreate, RequestOut

router = APIRouter(prefix="/requests", tags=["requests"])


@router.post("", response_model=RequestOut, status_code=status.HTTP_201_CREATED)
def submit_request(
    payload: RequestCreate,
    db: Session = Depends(get_db),
    current_employee: Employee = Depends(get_current_employee),
):
    if not payload.service_ids:
        raise HTTPException(status_code=400, detail="At least one service is required")

    services = db.query(Service).filter(Service.id.in_(payload.service_ids)).all()
    if len(services) != len(payload.service_ids):
        raise HTTPException(status_code=404, detail="One or more services not found")

    request = Request(employee_id=current_employee.id, type=payload.type, total_all=0)
    db.add(request)
    db.flush()

    total_all = 0
    for service in services:
        item = RequestItem(
            request_id=request.id,
            service_id=service.id,
            title=service.title,
            provider_name=service.provider_name,
            original_price_all=service.price_all,
            discounted_price_all=service.price_all,
        )
        db.add(item)
        db.add(ActiveService(employee_id=current_employee.id, request_id=request.id, service_id=service.id))
        total_all += service.price_all

    request.total_all = total_all
    db.commit()
    db.refresh(request)
    return request


@router.get("", response_model=list[RequestOut])
def list_requests(
    status_filter: str | None = None,
    db: Session = Depends(get_db),
    current_manager: Employee = Depends(get_current_manager),
):
    query = db.query(Request)
    if status_filter:
        query = query.filter(Request.status == status_filter)
    return query.all()


@router.post("/{request_id}/approve", response_model=RequestOut)
def approve_request(
    request_id: int,
    payload: RequestApprove,
    db: Session = Depends(get_db),
    current_manager: Employee = Depends(get_current_manager),
):
    request = db.get(Request, request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Request not found")

    request.status = "approved"
    request.paid_at = datetime.now(timezone.utc)
    request.payment_method = payload.payment_method

    for active_service in db.query(ActiveService).filter(ActiveService.request_id == request.id):
        active_service.status = "active"

    for item in request.items:
        db.add(
            RedeemedHistory(
                employee_id=request.employee_id,
                service_id=item.service_id,
                title=item.title,
                provider_name=item.provider_name,
                price_all=item.discounted_price_all,
                claimed_at=request.paid_at,
            )
        )

    db.commit()
    db.refresh(request)
    return request


@router.post("/{request_id}/decline", response_model=RequestOut)
def decline_request(
    request_id: int,
    db: Session = Depends(get_db),
    current_manager: Employee = Depends(get_current_manager),
):
    request = db.get(Request, request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Request not found")

    request.status = "declined"
    db.query(ActiveService).filter(ActiveService.request_id == request.id).delete()

    db.commit()
    db.refresh(request)
    return request
